### PR TITLE
K8SPSMDB-268: TLS fix

### PR DESCRIPTION
**Pull request by qJkee**
[K8SPSMDB-268](https://jira.percona.com/browse/K8SPSMDB-268)

When using certmanager, we can not create cluster successfully because of certificates errors

### CODE REVIEW

**Comment by toabi** [comment]
Nice. Noticed the missing apiGroup recently too. Would be even more awesome if it would work with 1.0+ :D

**Comment by qJkee** [comment]
> Nice. Noticed the missing apiGroup recently too. Would be even more awesome if it would work with 1.0+ :D

Hey Tobias!
I'm fixing the final issue with tls, since somehow mongodb rejects certificates provided by certmanager
Doing all my best =)

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-434-aac3804a

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| init-deploy | passed |
| storage | passed |
| upgrade | passed |
| monitoring | passed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| self-healing | passed |
| security-context | passed |
| scheduled-backup | passed |
| arbiter | passed |
| operator-self-healing | passed |
| service-per-pod | passed |
| liveness | passed |
| users | passed |

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-434-c55095e5

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| storage | passed |
| init-deploy | passed |
| upgrade | passed |
| self-healing | passed |
| monitoring | passed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| operator-self-healing | passed |
| security-context | passed |
| scheduled-backup | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| users | passed |

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-434-07c333f1

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| init-deploy | passed |
| storage | passed |
| upgrade | passed |
| monitoring | passed |
| self-healing | passed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| security-context | passed |
| operator-self-healing | passed |
| scheduled-backup | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| users | passed |

File: pkg/psmdb/mongo/mongo.go [hunk 1]
```diff
@@ -32,11 +32,6 @@ func Dial(addrs []string, replset, username, password string, useTLS bool) (*mon
 		SetWriteConcern(writeconcern.New(writeconcern.WMajority(), writeconcern.J(true))).
 		SetReadPreference(readpref.Primary())
 
-	if useTLS {
-		tlsCfg := tls.Config{InsecureSkipVerify: true}
-		opts = opts.SetTLSConfig(&tlsCfg).SetDialer(tlsDialer{cfg: &tlsCfg})
-	}
-
```
Comment (delgod) [inline pkg/psmdb/mongo/mongo.go:39]: I believe if allowUnsafeConfiguration=false we should fetch CA from secret and use TLS connection between Operator and MongoDB

**Review by delgod — CHANGES_REQUESTED** [review]

File: pkg/controller/perconaservermongodb/version.go [hunk 2]
```diff
@@ -188,8 +189,16 @@ func (r *ReconcilePerconaServerMongoDB) fetchVersionFromMongo(cr *api.PerconaSer
...
-	session, err := mongo.Dial(rsAddrs, replset.Name, string(usersSecret.Data[envMongoDBClusterAdminUser]), string(usersSecret.Data[envMongoDBClusterAdminPassword]), false)
+	username := string(usersSecret.Data[envMongoDBClusterAdminUser])
+	password := string(usersSecret.Data[envMongoDBClusterAdminPassword])
+	var tlsConf *tls.Config
+	if !cr.Spec.UnsafeConf {
+		tlsConf, err = r.mongoTLSConf(cr.Spec.Secrets.SSL, cr.Namespace)
+		if err != nil {
+			return errors.Wrap(err, "get mongo tls config")
+		}
+	}
```
Comment (delgod) [inline pkg/controller/perconaservermongodb/version.go:200]: why not move this inside `Dial` function?

**Review by delgod — CHANGES_REQUESTED** [review]

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-434-de187194

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | failed |
| storage | failed |
| upgrade | failed |
| one-pod | failed |
| upgrade-consistency | failed |
| limits | failed |
| monitoring | failed |
| self-healing | failed |
| demand-backup | failed |
| scaling | failed |
| operator-self-healing | failed |
| monitoring-2-0 | failed |
| scheduled-backup | failed |
| security-context | failed |
| arbiter | failed |
| service-per-pod | failed |
| liveness | failed |
| users | failed |

**Review by hors — CHANGES_REQUESTED** [review]
@qJkee  please add simple DB connection check into init-deploy test. We need to get the number of DB connections and the value should not be more than 5.

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | passed |
| one-pod | passed |
| upgrade | passed |
| self-healing | passed |
| monitoring | passed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| security-context | passed |
| operator-self-healing | passed |
| scheduled-backup | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| users | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| storage | passed |
| init-deploy | passed |
| upgrade | passed |
| self-healing | passed |
| monitoring | passed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| demand-backup | passed |
| monitoring-2-0 | passed |
| security-context | passed |
| operator-self-healing | passed |
| arbiter | passed |
| scheduled-backup | passed |
| service-per-pod | passed |
| liveness | passed |
| users | failed |

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-434-fc78df9a

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| storage | passed |
| init-deploy | passed |
| upgrade | passed |
| monitoring | passed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| self-healing | passed |
| security-context | passed |
| scheduled-backup | passed |
| arbiter | passed |
| operator-self-healing | passed |
| service-per-pod | passed |
| liveness | passed |
| users | passed |

**Review by delgod — APPROVED** [review]

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-434-eeb4d809

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | passed |
| one-pod | passed |
| upgrade | passed |
| self-healing | passed |
| monitoring | passed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| security-context | passed |
| operator-self-healing | passed |
| scheduled-backup | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| users | passed |

**Review by hors — APPROVED** [review]

**Review by blez — APPROVED** [review]

**Pull request merged**